File: backend/src/controllers/StateController.php
<?php

namespace App\Controllers;

use App\Controllers\Traits\JsonResponseTrait;
use App\Models\State;
use Psr\Http\Message\ResponseInterface as Response;
use Psr\Http\Message\ServerRequestInterface as Request;

class StateController
{
    use JsonResponseTrait;

    /**
     * GET /states
     */
    public function getAll(Request $request, Response $response): Response
    {
        try {
            $states = State::all();
            return $this->jsonResponse($response, 200, true, "All states fetched", $states);
        } catch (\Exception $e) {
            return $this->jsonResponse($response, 500, false, "Error fetching states", [
                "details" => $e->getMessage()
            ]);
        }
    }

    /**
     * GET /states/{id}
     */
    public function getById(Request $request, Response $response, array $args): Response
    {
        try {
            $state = State::find($args['id']);
            if (!$state) {
                return $this->jsonResponse($response, 404, false, "State not found");
            }
            return $this->jsonResponse($response, 200, true, "State fetched successfully", $state);
        } catch (\Exception $e) {
            return $this->jsonResponse($response, 500, false, "Error fetching state", [
                "details" => $e->getMessage()
            ]);
        }
    }

    /**
     * POST /states
     */
    public function create(Request $request, Response $response): Response
    {
        try {
            $data = json_decode($request->getBody(), true);

            if (empty($data['StateName'])) {
                return $this->jsonResponse($response, 400, false, "Missing required field: StateName");
            }

            $state = State::create($data);
            return $this->jsonResponse($response, 201, true, "State created successfully", $state);
        } catch (\Exception $e) {
            return $this->jsonResponse($response, 500, false, "Error creating state", [
                "details" => $e->getMessage()
            ]);
        }
    }

    /**
     * PUT /states/{id}
     */
    public function update(Request $request, Response $response, array $args): Response
    {
        try {
            $state = State::find($args['id']);
            if (!$state) {
                return $this->jsonResponse($response, 404, false, "State not found");
            }

            $data = json_decode($request->getBody(), true);
            $state->update($data);

            return $this->jsonResponse($response, 200, true, "State updated successfully", $state);
        } catch (\Exception $e) {
            return $this->jsonResponse($response, 500, false, "Error updating state", [
                "details" => $e->getMessage()
            ]);
        }
    }

    /**
     * DELETE /states/{id}
     */
    public function delete(Request $request, Response $response, array $args): Response
    {
        try {
            $state = State::find($args['id']);
            if (!$state) {
                return $this->jsonResponse($response, 404, false, "State not found");
            }
            $state->delete();

            return $this->jsonResponse($response, 204, true, "State deleted successfully");
        } catch (\Exception $e) {
            return $this->jsonResponse($response, 500, false, "Error deleting state", [
                "details" => $e->getMessage()
            ]);
        }
    }
}

File: backend/src/main/java/com/nearchrist/backend/controller/StateController.java
package com.nearchrist.backend.controller;

import com.nearchrist.backend.dto.StateDto;
import com.nearchrist.backend.dto.StateUpsertDto;
import com.nearchrist.backend.dto.ApiResponse;
import com.nearchrist.backend.service.StateService;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/states")
public class StateController {
    private final StateService service;

    public StateController(StateService service) {
        this.service = service;
    }

    @GetMapping
    public ResponseEntity<ApiResponse<List<StateDto>>> getAll() {
        try {
            List<StateDto> states = service.getAllStates();
            return ResponseEntity.ok(new ApiResponse<>(true, 200, "All states fetched", states));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(new ApiResponse<>(false, 500, "Error fetching states: " + e.getMessage(), null));
        }
    }

    @GetMapping("/{id}")
    public ResponseEntity<ApiResponse<StateDto>> getById(@PathVariable Long id) {
        try {
            return service.getStateById(id)
                    .map(s -> ResponseEntity.ok(new ApiResponse<>(true, 200, "State fetched successfully", s)))
                    .orElseGet(() -> ResponseEntity.status(HttpStatus.NOT_FOUND)
                            .body(new ApiResponse<>(false, 404, "State not found", null)));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(new ApiResponse<>(false, 500, "Error fetching state: " + e.getMessage(), null));
        }
    }

    @PostMapping
    public ResponseEntity<ApiResponse<StateDto>> create(@RequestBody StateUpsertDto dto) {
        try {
            StateDto saved = service.createState(dto);
            return ResponseEntity.status(HttpStatus.CREATED)
                    .body(new ApiResponse<>(true, 201, "State created successfully", saved));
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest()
                    .body(new ApiResponse<>(false, 400, e.getMessage(), null));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(new ApiResponse<>(false, 500, "Error creating state: " + e.getMessage(), null));
        }
    }

    @PutMapping("/{id}")
    public ResponseEntity<ApiResponse<StateDto>> update(@PathVariable Long id, @RequestBody StateUpsertDto dto) {
        try {
            return service.updateState(id, dto)
                    .map(s -> ResponseEntity.ok(new ApiResponse<>(true, 200, "State updated successfully", s)))
                    .orElseGet(() -> ResponseEntity.status(HttpStatus.NOT_FOUND)
                            .body(new ApiResponse<>(false, 404, "State not found", null)));
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest()
                    .body(new ApiResponse<>(false, 400, e.getMessage(), null));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(new ApiResponse<>(false, 500, "Error updating state: " + e.getMessage(), null));
        }
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<ApiResponse<Void>> delete(@PathVariable Long id) {
        try {
            if (!service.deleteState(id)) {
                return ResponseEntity.status(HttpStatus.NOT_FOUND)
                        .body(new ApiResponse<>(false, 404, "State not found", null));
            }
            return ResponseEntity.ok(new ApiResponse<>(true, 204, "State deleted successfully", null));
        } catch (IllegalStateException e) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                    .body(new ApiResponse<>(false, 400, e.getMessage(), null));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(new ApiResponse<>(false, 500, "Error deleting state: " + e.getMessage(), null));
        }
    }
}